language/vm/vm_runtime/src/data_cache.rs

```diff
@@ -246,9 +246,27 @@ impl<'txn> TransactionDataCache<'txn> {
                 let deleted = global_ref.is_deleted();
                 if let Some(data) = global_ref.get_data() {
                     if deleted {
-                        write_set.push((key, WriteOp::Deletion));
+                        if write_set.len().checked_add(1).is_some() {
+                            // Implication of checked add
+                            verify!(write_set.len() < usize::max_value());
+                            write_set.push((key, WriteOp::Deletion));
+                        } else {
+                            return Err(VMRuntimeError {
+                                loc: Location::new(),
+                                err: VMErrorKind::DataFormatError,
+                            });
+                        }
                     } else if let Some(blob) = data.simple_serialize() {
-                        write_set.push((key, WriteOp::Value(blob)));
+                        if write_set.len().checked_add(1).is_some() {
+                            // Implication of checked add
```

Comment:
Why not just check for the implication? We don't actually use the result of the checked_add.